File: backend/internal/entity/flight.go
package entity

import (
	"fmt"

	"cloud.google.com/go/civil"
)

type Flight struct {
	Legs []FlightLeg `json:"legs"`
}

type Airport struct {
	Iata         string   `json:"iata"`
	Name         string   `json:"name"`
	Municipality string   `json:"municipality"`
	Location     Location `json:"location"`
}

type AirportWithTimezone struct {
	Airport
	Timezone string `json:"timezone"`
}

type FlightLeg struct {
	Origin            Airport        `json:"origin"`
	Destination       Airport        `json:"destination"`
	Airline           string         `json:"airline"`
	FlightNumber      string         `json:"flightNumber"`
	DepartureDateTime civil.DateTime `json:"departureDateTime"`
	ArrivalDateTime   civil.DateTime `json:"arrivalDateTime"`
	AmadeusFlightDate *civil.Date    `json:"amadeusFlightDate" extensions:"nullable"`
	DurationInMinutes int32          `json:"durationInMinutes"`
	Aircraft          *string        `json:"aircraft" extensions:"nullable"`
}

type PNR struct {
	Airline string `json:"airline" example:"LH"`
	PNR     string `json:"pnr"     example:"123456"`
}

type ErrAmbiguousFlightRequest map[string][]AmbiguousFlightChoice

type AmbiguousFlightChoice struct {
	OriginIata        string         `json:"originIata"`
	DestinationIata   string         `json:"destinationIata"`
	DepartureDateTime civil.DateTime `json:"departureDateTime"`
}

func (e ErrAmbiguousFlightRequest) Error() string {
	return fmt.Sprint(e)
}

File: app/src/domain/flight.ts
import * as z from "zod"
import { GeoJSONSchema } from "zod-geojson"
import { Location } from "./"

const Airport = z.object({
  iata: z.string(),
  name: z.string(),
  municipality: z.string(),
  location: Location,
})
export type Airport = z.infer<typeof Airport>

const FlightLeg = z.object({
  id: z.string(),
  origin: Airport,
  destination: Airport,
  airline: z.string(),
  flightNumber: z.string(),
  departureDateTime: z.iso.datetime(),
  arrivalDateTime: z.iso.datetime(),
  amadeusFlightDate: z.iso.date().optional(),
  durationInMinutes: z.number(),
  aircraft: z.string().optional(),
})
export type FlightLeg = z.infer<typeof FlightLeg>
const CreateFlightLeg = FlightLeg.omit({ id: true })

const PNR = z.object({
  id: z.string(),
  airline: z.string(),
  pnr: z.string(),
})
export type PNR = z.infer<typeof PNR>
const CreatePNR = PNR.omit({ id: true })

export const Flight = z.object({
  id: z.string(),
  type: z.literal("flight"),
  legs: z.array(FlightLeg),
  pnrs: z.array(PNR),
  price: z.number().optional(),
  geoJson: GeoJSONSchema.optional(),
})
export type Flight = z.infer<typeof Flight>

export const CreateFlight = z
  .object({
    ...Flight.shape,
    legs: z.array(CreateFlightLeg),
    pnrs: z.array(CreatePNR),
  })
  .omit({ id: true, type: true })
export type CreateFlight = z.infer<typeof CreateFlight>

export const UpdateFlight = CreateFlight.partial()
export type UpdateFlight = z.infer<typeof UpdateFlight>
